chore: field names must not be verbs. (#1078)
googleapis/aip: aip/general/0140.md

- `collected_items` (**not** `items_collected`)
- `imported_objects` (**not** `objects_imported`)

### Verbs

Field names **must not** be named to reflect an intent or action. They
**must not** be verbs. Rather, because the field defines the *desired

> value* for mutations, e.g. Create and Update, and the *current value*

Reviewer: I think the markdown for italicizing (at least in this "preview" on github) indicates the formatting isn't being applied correctly. Can you just double check that? Might need to close it on line 94 and reopen it on 95.

Author: Looks right on the [preview](https://github.com/aip-dev/google.aip.dev/pull/1078/files?short_path=7de9833#diff-7de983345e17e37eb41b1a8d22e476a53d0c0736fd074b7251341bbda2cb08f7), but I'll fire it up on site-generator just the same.

Reviewer: Ah ok cool. In code review mode, GH tries to get clever and semi-format the markdown, but it doesn't work correctly
![Screenshot 2023-04-27 at 9 50 27 AM](https://user-images.githubusercontent.com/6644735/234933947-a70fccf8-861c-4733-b786-9f7401ff75e7.png)

See how `value*` isn't italicized, but the following text is, incorrectly? Boo GitHub.

Author: Yeah, it's different on src vs preview. :(

Author: Verified that it renders correctly with local dev env of site-generator...

for reads, e.g.  Get and List, the name **must** be a noun. It defines
what is so, not what to do.

- `collected_items` (**not** `collect_items`)
- `disabled` (**not** `disable`)

In contrast, method names, whether standard or custom, change facets of

> resources and are be named as verbs.

Reviewer 2: Choose one verb :-)

Author: :-)

- `collected_items` (**not** `items_collected`)
- `imported_objects` (**not** `objects_imported`)

### Verbs

Field names **must not** be expressed to reflect an intent or action.
That is, they **must not** be expressed as verbs. Rather, the field name
should express the condition of that aspect of the resource as defined
by the value of the field in question.

> When performing a mutation, the field value is set to the desired

Reviewer 3: this feels like something that belongs in "rationale" to me, but it's not required.

Author: I see your point, **and** I like it inline. Let me see if I can reword so that it feels more appropriate where it is.

*final* value of that property. When performing a read, the field value
is returned as the *current* value of that property. As such, the name
should always be *a noun or an adjective* as it describes what is so,
not what to do.

- `collected_items` (**not** `collect_items`)

> - `imported_objects` (**not** `import_objects`)

Reviewer: Can we include a boolean example? These are by far the worst culprits in API review
```suggestion
- `imported_objects` (**not** `import_objects`)
- `extra_logging_enabled` (**not** `enable_extra_logging`)
```

Author: Agreed, good one! Fixing.

Alternatively, method names, whether standard or custom, express intent

> to change the condition of resources and **should** be named as verbs,

Reviewer: I think we should avoid guidance on Methods in this Field-specific AIP. As such I would reword this as a statement of fact.
```suggestion
to change the condition of resources and are named as verbs,
```

Author: Fixing.

accordingly.

> **Note:** Fields named as verbs may suggest that there should be a

Reviewer: ```suggestion
**Note:** Fields named as verbs could suggest that there is a
```
Nit: avoid using a requirement words :)

Author: Fixing.

Author: Ended up removing the whole note.

complementary, corresponding field that is a noun or adjective, one to

> set a value and the other to get the value, respectively. This is highly

Reviewer 3: I actually think this is the pattern we plan on recommending. @bgrant0607 for thoughts.

But I think you're referring to the idea of the input field being a verb and the output field being a noun or an adjective? which I agree isn't good.

I feel like you don't even need this paragraph. Your above explanation for how resources represent end state and therefore shouldn't include nouns is convincing as-is.

Author: > I actually think this is the pattern we plan on recommending

My understanding was that this was to be recommended iff the field had a default or normalized value. @bgrant0607, wdyt?

> But I think you're referring to the idea of the input field being a verb and the output field being a noun or an adjective? which I agree isn't good.

Not quite. What I am getting at is that,m

1. There should be only one field
2. It should not be an verb/action

If there needs to be two fields, one for input and output, something has been implemented incorrectly, except in rare cases, e.g. when values **must** be defaulted or normalized. Then, it may be correct to have an OUTPUT_ONLY counterpart to the input field.

> I feel like you don't even need this paragraph. Your above explanation for how resources represent end state and therefore shouldn't include nouns is convincing as-is.

In the end, I think you are right. Will remove.

Reviewer 2: I'm not sure I understand the question.

My canonical example is the Kubernetes Deployment spec "paused" field. "paused" not "pause". That matches this proposal, I believe.

The "double simplex" approach we've been discussing to address static defaults, semi-static defaults, configured defaults, dynamic defaults, semantic strings, associative lists, value normalizations (e.g., lower casing or upper casing), value additions, and other service-side transformations is a different issue, I believe. 

Example: `foo` and output-only `effective_foo`.

Unless this is a reference to the desired_state proposal or similar case where the service may want to return status information regarding whether the desired state has been achieved.

Author: @bgrant0607 I think we were asking more about the "double simplex" approach and I think you answered it. We were discussing the paragraph (removed) about that it is a poor practice to have two fields--one with a verb name and one with a noun name--because it is not IaC friendly. However, we now realized that this is a good pattern for the cases that you mentioned. But, as I said, this paragraph was removed and we appreciate your input here.

> discouraged, as it is suboptimal for declarative client use cases.

Reviewer 3: ```suggestion
discouraged, as it is suboptimal for IaC client use cases.
```

We defined the term IaC: https://google.aip.dev/9#iac. Although perhaps we might want to rename the term to declarative.

Author: Removing entire paragraph anyway. Resolving.
